### DISCUSSION TITLE
Can you define the line ending type for generated makefiles?

### DISCUSSION THREAD

**Discussion by ronsonm** [eclipse-cdt/cdt#232] [Q&A]
Question:
Is it possible to define which type of line endings style (UNIX LF / Windows CRLF) are used inside the generated makefiles from CDT?

Background info / explanation:
I have an eclipse project checked out on a windows machine with C/C++CDT configured to generate GNU makefiles. The generated makefiles use a mixture of LF and CRLF line endings. I want to force the makefile generator to use UNIX (LF) line endings. Is this possible? Note: I am aware of the "Window -> Preferences -> Workspace -> New text file line delimiter -> Other -> Unix" setting, this however does not impact generated files and thus did not solve my issue.

**Comment by jantje** [comment]
I'm not 100% sure but I think Managed Build system uses the formatting of the existing files.

**Reply by ronsonm** [reply]
What existing files would that be? The makefiles are generated, the source code and other project files are formatted in the UNIX line endings style. It looks like the build system just uses the system's line endings.

**Reply by jantje** [reply]
It is actually up to the makefile generator to save the files.
The gnumakefilegenerator Does this here
https://github.com/eclipse-cdt/cdt/blob/main/build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/makegen/gnu/GnuMakefileGenerator.java#L814
Hope this helps

**Reply by ronsonm** [reply]
Thank you for that pointer to the relevant code. Although I don't believe the file encoding is what determines the line endings.
I glanced over that code and it seems to me that the line endings are defined by a `NEWLINE` definition (like [here](https://github.com/eclipse-cdt/cdt/blob/main/build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/makegen/gnu/GnuMakefileGenerator.java#L2043)). I can't however determine where this definition comes from and how one could force CDT to set this to LF instead of CRLF.

**Reply by jantje** [reply]
I'm not a java programmer ;-)
Anyways seems your perception on system level selection is right.
https://github.com/eclipse-cdt/cdt/blob/main/build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/makegen/IManagedBuilderMakefileGenerator.java#L55

**Comment by jonahgraham** [comment]
> Is it possible to define which type of line endings style (UNIX LF / Windows CRLF) are used inside the generated makefiles from CDT?

AFAIK it is not possible as the newlines are hardcoded. 

The newline is normally system dependent:

https://github.com/eclipse-cdt/cdt/blob/3feeedfa63ac4a01fca911a78376567c7d3c2bbb/build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/makegen/IManagedBuilderMakefileGenerator.java#L55

But escaped newlines are handled differently: 

https://github.com/eclipse-cdt/cdt/blob/3feeedfa63ac4a01fca911a78376567c7d3c2bbb/build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/makegen/IManagedBuilderMakefileGenerator.java#L45

That `\\\n` pattern is used to generate the end of this line in sources.mk (amongst other places)
```
# Every subdirectory with source files must be described here
SUBDIRS := \
```

On a quick read of this code, since we sometimes use `\n` I don't know why we then sometimes use `\r\n` on Windows. I guess changing it all to `\n` all the time would be acceptable. I don't know what it would take to verify that change in the code base, nor if it would be worth it. Feel free to continue the conversation if this is important enough to you to spend some time providing a PR. 

PS the `.d` files which are included by the Makefiles are generated by GCC, so the newlines in them may be different.
